### DISCUSSION TITLE
rendering an stl file on the click of a button

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [zauberzeug/nicegui#1203] [Q&A] [answered]
### Question

Just started https://github.com/WolfgangFahl/nicescad to try to avoid to have to use https://github.com/nickc92/ViewSCAD and running a jupyter notebook. I was hoping to get  a simple rendering done with nicegui.

The code i am trying is here:
https://github.com/WolfgangFahl/nicescad/blob/f52b1561f525c93f32950dafd3c9c72a7fefa1cc/nicescad/webserver.py#L34

When i click the render button i get the error message:
```python
    self.scene.stl("/stl/tmp.stl").move(x=-0.5).scale(0.06)
  File "/Users/wf/Library/Python/3.10/lib/python/site-packages/nicegui/elements/scene_objects.py", line 101, in __init__
    super().__init__('stl', url, wireframe)
  File "/Users/wf/Library/Python/3.10/lib/python/site-packages/nicegui/elements/scene_object3d.py", line 18, in __init__
    self.scene.objects[self.id] = self
AttributeError: 'Splitter' object has no attribute 'objects'
```

The culprit seems to be
```python
       self.scene: 'Scene' = cast('Scene', globals.get_slot().parent)
        self.scene.objects[self.id] = self
```

i do not understand the cast to Scene for the parent here. What's wrong with the approach i am taking?

**Comment by falkoschindler** [comment, marked as answer]
Hi Wolfgang,

When creating 3D objects within a `ui.scene`, you're using something like `scene.sphere()`. But `.sphere()` is static, thus doesn't know _which_ scene it belongs to. Therefore you need to enter the context of the scene element before creating objects. So your code would look something like this:
```py
with self.scene:
    self.scene.stl("/stl/tmp.stl").move(x=-0.5).scale(0.06)
```

This looks a bit weird, I have to admit. We usually import the class of the 3D object and create it like this (see [here](https://github.com/zauberzeug/rosys/blob/main/rosys/pathplanning/obstacle_object_.py) for a real-life example):
```py
with self.scene:
    Sphere()
```

But your style is actually more in line with how regular NiceGUI elements are created. Maybe we find a way to support constructing objects via `scene.sphere()` without the need for setting an explicit context `with scene`. I'll think about it.

Regarding the `cast` you've been referring to: The object holds a reference to its scene. And since the current context should be a scene, `globals.get_slot().parent` can be safely casted to `Scene`. But this is only for static type checkers like mypy. The Python runtime itself doesn't care about the `cast()` function. Therefore the `scene` is set to whatever the parent of the current slot is (a `Splitter` in this case). Only later, when accessing the scene's `objects`, an AttributeError occurs, since a `Splitter` has no `objects`.

**Comment by WolfgangFahl** [comment]
@falkoschindler  thanks this now at least renders correctly.
